src/codegen/intermediate_representation.c

```diff
@@ -645,14 +645,15 @@ IRInstruction *ir_create_static(CodegenContext *context, Type *type, span name)
   return ref;
 }
 
+/// NOTE: Currently unused, but can be used to load a static reference
+/// more than once without generating duplicate static variables.
 IRInstruction *ir_static_reference(CodegenContext *context, span name) {
   foreach_ptr(IRStaticVariable *, v, context->static_vars) {
     if (string_eq(v->name, name)) {
       INSTRUCTION(ref, IR_STATIC_REF);
       ref->static_ref = v;
       ref->type = ast_make_type_pointer(context->ast, v->type->source_location, v->type);
-      // TODO: `v->reference` may need to become list of references? I think this is why
-      // optimisation is broken.
+      // TODO: `v->reference` may need to become list of references?
```

Comment:
Well the original idea was that there would ever only be one IR_STATIC_REF for each static variable. That’s what there simply was no `ir_create_static_ref()`, simply because the idea of such an operation would have been flawed to begin with, at least when I wrote this.

Reply:
Interesting; so there's only some confusion in the fact that IR_STATIC_REF both creates a static as well as holds a reference to an existing static (it's address), in the same way that ALLOCA does. However, can't a static be referenced from multiple scopes? If so, won't each one need a reference to the static? Maybe I'm thinking about this the wrong way

Reply:
When you create a static variable, it creates an `IRStaticVariable`, which is added to a vector in the context and it’s how the variables are emitted by the backend. This structure holds information about e.g. the name of a static variable:
https://github.com/LensPlaysGames/FUNCompiler/blob/8282edc50843fc5ec882469198e1aedb149dd46e/src/codegen.h#L40-L44

By contrast, an `IR_STATIC_REF` is an instruction that wraps an `IRStaticVariable`, in the same way that an `IR_ALLOCA` wraps an `IRStackAllocation`. The `IRStaticVariable` *is* the variable, the IR_STATIC_REF contains a pointer to that variable. The type of an `IR_STATIC_REF` is a pointer to whatever the type of the static variable is, for the same reason that the type of an `IR_ALLOCA` is a pointer to the allocated storage.

The `IR_STATIC_REF` is just a way to refer to an `IRStaticVariable` in IR. Only one is needed per variable as just like the `IR_STATIC_REF` holds a pointer to the `IRStaticVariable`, so does the `IRStaticVariable` contain a pointer to its `IR_STATIC_REF`. This is why the (not ‘a’) `IR_STATIC_REF` for a static variable is created together with that variable